src/pulselistener/pulselistener/mercurial.py

```diff
@@ -19,93 +20,263 @@
+
+            message = ''
+            if commit:
+                message += '{}\n'.format(commit[0]['message'])
+            message += 'Differential Diff: {}'.format(diff_phid)
+
+            logger.info('Applying patch', phid=diff_phid, message=message)
+            self.repo.import_(
+                patches=io.BytesIO(patch.encode('utf-8')),
+                message=message,
+                user='pulselistener',
+            )
+            await asyncio.sleep(1)
+
+    def add_try_commit(self, build):
+        '''
+        Build and commit the file configuring try
+        * try_task_config.json in json mode
+        * .try in syntax mode
+        '''
+        if self.try_mode == TryMode.json:
+            path = os.path.join(self.dir, 'try_task_config.json')
+            config = {
+                'version': 2,
+                'parameters': {
+                    'target_tasks_method': 'codereview',
+                    'optimize_target_tasks': True,
+                    'phabricator_diff': build.target_phid,
+                }
+            }
+            content = json.dumps(config, sort_keys=True, indent=4)
+            message = 'try_task_config for code-review\nDifferential Diff: {}'.format(build.diff['phid'])
+
+        elif self.try_mode == TryMode.syntax:
+            path = os.path.join(self.dir, '.try')
+            content = 'try: {}'.format(self.try_syntax)
+            message = content
+
+        else:
+            raise Exception('Unsupported try mode')
+
+        # Write content and commit it
+        with open(path, 'w') as f:
+            f.write(content)
+        self.repo.add(path.encode('utf-8'))
+        self.repo.commit(
+            message=message,
+            user='pulselistener',
+        )
+
+    async def push_to_try(self):
+        '''
+        Push the current tip on remote try repository
+        '''
+        tip = self.repo.tip()
+        logger.info('Pushing patches to try', rev=tip.node)
+        self.repo.push(
+            dest=self.try_url.encode('utf-8'),
+            rev=tip.node,
+            ssh=self.ssh_conf,
+            force=True,
+        )
+        return tip
+
+    def clean(self):
+        '''
+        Steps to clean the mercurial repo
+        '''
+        logger.info('Remove uncommited changes')
+        self.repo.revert(self.dir.encode('utf-8'), all=True)
+
+        logger.info('Remove all mercurial drafts')
+        try:
+            cmd = hglib.util.cmdbuilder(b'strip', rev=b'roots(outgoing())', force=True, backup=False)
+            self.repo.rawcommand(cmd)
+        except hglib.error.CommandError as e:
+            if b'abort: empty revision set' not in e.err:
+                raise
+
+        logger.info('Pull updates from remote repo')
+        self.repo.pull()
+
+
+class MercurialWorker(object):
+    '''
+    Mercurial worker maintaining a local clone of mozilla-unified
+    '''
+    def __init__(self, phabricator_api, publish_phabricator, cache_root, repositories):
```

Comment:
I'm not a fan of moving the Phab stuff into MercurialWorker, as we are mixing two things that are logically different, but I'm OK with it if it's too much work to keep them separate.

Reply:
This is a nice refactoring i'd like to do. Filed #2167 to track this